File: benchmarks/tsb/bench_multi_index_contains.ts
/**
 * Benchmark: MultiIndex.contains — check if a tuple key exists in the index.
 */
import { MultiIndex } from "../../src/index.js";

const SIZE = 5_000;
const WARMUP = 5;
const ITERATIONS = 50;

const arr1 = Array.from({ length: SIZE }, (_, i) => `a${i % 50}`);
const arr2 = Array.from({ length: SIZE }, (_, i) => i % 100);
const mi = MultiIndex.fromArrays([arr1, arr2]);

for (let i = 0; i < WARMUP; i++) {
  mi.contains(["a0", 0]);
}

const t0 = performance.now();
for (let i = 0; i < ITERATIONS; i++) {
  mi.contains([`a${i % 50}`, i % 100]);
}
const total = performance.now() - t0;

console.log(
  JSON.stringify({ function: "multi_index_contains", mean_ms: total / ITERATIONS, iterations: ITERATIONS, total_ms: total }),
);

File: benchmarks/pandas/bench_multi_index_contains.py
"""Benchmark: MultiIndex.__contains__ (pandas equivalent)."""
import json
import time
import pandas as pd

SIZE = 5_000
WARMUP = 5
ITERATIONS = 50

arr1 = [f"a{i % 50}" for i in range(SIZE)]
arr2 = [i % 100 for i in range(SIZE)]
mi = pd.MultiIndex.from_arrays([arr1, arr2])

for _ in range(WARMUP):
    ("a0", 0) in mi

t0 = time.perf_counter()
for i in range(ITERATIONS):
    (f"a{i % 50}", i % 100) in mi
total = (time.perf_counter() - t0) * 1000

print(json.dumps({"function": "multi_index_contains", "mean_ms": total / ITERATIONS, "iterations": ITERATIONS, "total_ms": total}))
